adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+| [code]#work-group barrier#                                         | [code]#__syncthread#
+|====
+
+[[sec::programming_interface]]
+== Programming Interface
+
+[[sub:cuda:queries]]
+=== Queries
+
+For all event information profiling descriptors, the calls to 
+[code]#sycl::event::get_profiling_info# return the time (in nanoseconds) since 
+the creation of the context that the event is associated with. The "Resolution"
+(timing error) of the returned value is the same as that provided by the CUDA
+driver API call `cuEventElapsedTime`: +/- 0.5 microseconds. All event
+information profiling descriptors, defined by the SYCL specification, are
+supported by the CUDA backend.
+
+Currently no restrictions are defined for parameters of [code]#get_info# member
+function in classes [code]#platform#, [code]#context#, [code]#device#, 
+[code]#queue#, [code]#event# and [code]#kernel#. All parameter values defined 
+in the SYCL specification are supported.
+
+Querying for [code]#info::device::backend_version# by calling 
+[code]#device::get_info# returns the CUDA compute capability of the device.
+
+Currently no parameters are defined for [code]#get_backend_info# member 
+functions of classes [code]#platform#, [code]#context#, [code]#device#, 
+[code]#queue#, [code]#event# and [code]#kernel#.
+
+[[sub:cuda:application_interoperability]]
+=== Application Interoperability
+
+This section describes the API level interoperability between SYCL and CUDA.
+
+The CUDA backend supports API interoperability for `platform`, `device`,
+`context`, `queue`, and `event`. Interoperability for `buffer`, `kernel`,
+`kernel_bundle`, `device_image`, `sampled_image` and `unsampled_image` are not
+supported.
+
+[[table.cuda.appinterop.nativeobjects]]
+.Types of native backend objects application interoperability
+[width="100%",options="header",cols="20%,20%,20%,40%"]
+|====
+| [code]#SyclType# | [code]#backend_input_t<backend::cuda, SyclType># | [code]#backend_return_t<backend::cuda, SyclType># | Description
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts, however, it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue can encapsulates multiple CUDA stream, however, a SYCL queue can only be created from or produce one, and any synchronization required should be performed.
+| [code]#event#    | `CUevent`    | `CUevent`  | A SYCL event can encapsulates multiple CUDA events, however, a SYCL event can only be created from or produce one, and a CUevent produced from a SYCL event may or may not be valid, use `sycl::cuda::has_native_event` to query this.
```

Comment:
What happens if you call `get_native` for an `event` that  does not have a native event?

Same question here as above for `queue`.  What does it mean that a SYCL `event` could contain more than one native even but only one is returned from `get_native`?  Is it important for the application to know which is returned?

Reply:
cc @AerialMantis

Reply:
For OpenCL, `get_native<event>` returns a vector of `cl_event` objects.  Why wouldn't we do the same thing for CUDA and return a vector of `CUevent` objects?

Reply:
So that is definitely an option, we did think about having `get_native<event>` return all native events that are associated with the SYCL `event`, however, it introduces the problem of conveying to the user what each of these events represent in order for the user to utilise the individual events. Whereas we believe generally users just wish to just synchronize with all the commands associated with the `event`. We could expand this in the future to return all native events, but we thought that it was best to keep it simple to begin with until we have some more user experience with it. If you think this would be the better approach we could reconsider it.

Reply:
I also don't know what is easiest for users.  I just seems weird that we would decide one thing for OpenCL and a different thing for CUDA.  It will probably be difficult to change this later to return a vector of events because that would be an API-breaking change.